### PR TITLE
Fix type parsing issues in ILLink and ILC

### LINKED ISSUE #98955: Type.GetType("TypeName") will pick the "first" such type found
If we're resolving type name -> `TypeDefinition` (for example `Type.GetType` does this), and the type name is not assembly qualified, we will iterate over all assemblies and search each one for that type name. Once we find the first, we'll use that.

There's no guarantee it's the right one (if there are multiple). It would be easy to write a counter example.

### PR DESCRIPTION
**Pull request by sbomer**
- Fixes https://github.com/dotnet/runtime/issues/98955

   We will now produce a warning when a non-assembly-qualified type flows into a string location annotated with DynamicallyAccessedMembers, and we don't try to look up or mark the type (since we don't know which assemblies will be searched at runtime by the Type.GetType call).

- Fixes https://github.com/dotnet/runtime/issues/103906

   The ILLink intrinsic handling for `Type.GetType` will now look in corelib for generic arguments, matching native AOT.

This replaces the existing warning IL2105. I left that one alone instead of repurposing it, because we already documented IL2105 and older versions of ILLink will produce it. Best to avoid any confusion about them.

### CODE REVIEW

File: src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs [hunk 1]
```diff
@@ -52,8 +57,28 @@ public static class CustomAttributeTypeNameParser
...
             return type;
+
+            static bool IsFullyQualified (TypeName typeName)
+            {
+                if (typeName.AssemblyName is null)
+                {
+                    return false;
+                }
+
+                if (typeName.IsConstructedGenericType)
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs:69]: This does not seem to handle arrays/byrefs/pointers of generic types with unqualified argument.

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs:69]: I meant type name like `G`1[[MyType]][], MyAssembly`. I still think that type names like that are not handled correctly here.

(The tests you have added are for type names like `G`1[[MyType[]]], MyAssembly`. I agree that those are handled correctly.)

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/TrimmingTests/DebuggerTypeProxyAttributeTests.cs [hunk 2]
```diff
@@ -53,7 +53,7 @@ public DebuggerProxy(MyClass instance)
         }
     }
 
-    [DebuggerTypeProxy("Program+MyClassWithProxyStringProxy")]
+    [DebuggerTypeProxy("Program+MyClassWithProxyStringProxy, project")]
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/TrimmingTests/DebuggerTypeProxyAttributeTests.cs:56]: I think I'm fine with this behavior. I searched a few places that use Type.GetType currently, and they all appear to be using assembly qualified names. If we see new warnings, we should fix them.

File: src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs [hunk 3]
```diff
@@ -52,8 +57,33 @@ public static class CustomAttributeTypeNameParser
 
             TypeDesc type = resolver.Resolve(parsed);
 
-            typeWasNotFoundInAssemblyNorBaseLibrary = resolver._typeWasNotFoundInAssemblyNorBaseLibrary;
             return type;
+
+            static bool IsFullyQualified (TypeName typeName)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs:62]: We don't do this around here

```suggestion
            static bool IsFullyQualified(TypeName typeName)
```

File: src/tools/illink/src/linker/Linker.Steps/RootAssemblyInputStep.cs [hunk 4]
```diff
@@ -86,6 +86,9 @@ protected override void Process ()
 
 				// No metadata trimming
 				Context.MetadataTrimming = MetadataTrimming.None;
+
+				// Library mode builds against System.Runtime ref assembly
+				Context.SystemModuleName = "System.Runtime";
```
Comment (MichalStrehovsky) [inline src/tools/illink/src/linker/Linker.Steps/RootAssemblyInputStep.cs:91]: We should know what the core assembly is (it's the one that defines System.Object). Why is this extra configuration needed? I think I didn't fully understand https://github.com/dotnet/runtime/pull/104060#discussion_r1655659027.

**Review by MichalStrehovsky — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
@sbomer is it possible IL2122 might still appear even if `SuppressTrimAnalysisWarnings=true` / `--notrimwarn` was passed in?

image

We noticed this here:
* https://github.com/dotnet/android/pull/9108

**Pull request merged**